src/roles/_shaman_helper.py

```diff
@@ -0,0 +1,423 @@
+        if loser in evt.data["dead"] or victim is loser:
+            loser = None
+        if loser is not None:
+            prots = deque(var.ACTIVE_PROTECTIONS[loser.nick])
+            while len(prots) > 0:
+                # an event can read the current active protection and cancel the totem
+                # if it cancels, it is responsible for removing the protection from var.ACTIVE_PROTECTIONS
+                # so that it cannot be used again (if the protection is meant to be usable once-only)
+                ret_evt = Event("retribution_totem", {"message": []})
+                if not ret_evt.dispatch(var, victim, loser, prots[0]):
+                    evt.data["message"].extend(ret_evt.data["message"])
+                    return
+                prots.popleft()
+            evt.data["dead"].append(loser)
+            if var.ROLE_REVEAL in ("on", "team"):
+                role = get_reveal_role(loser)
+                an = "n" if role.startswith(("a", "e", "i", "o", "u")) else ""
+                evt.data["message"].append(messages["totem_death"].format(victim, loser, an, role))
+            else:
+                evt.data["message"].append(messages["totem_death_no_reveal"].format(victim, loser))
+
+@event_listener("transition_day_end", priority=1)
+def on_transition_day_end(evt, var):
+    message = []
+    for player, tlist in itertools.groupby(havetotem):
+        ntotems = len(list(tlist))
+        message.append(messages["totem_posession"].format(
+            player, "ed" if player not in get_players() else "s", "a" if ntotems == 1 else "\u0002{0}\u0002".format(ntotems), "s" if ntotems > 1 else ""))
+    for player in brokentotem:
+        message.append(messages["totem_broken"].format(player))
+    channels.Main.send("\n".join(message))
+
+@event_listener("begin_day")
+def on_begin_day(evt, var):
+    # Apply totem effects that need to begin on day proper
+    var.EXCHANGED.update(p.nick for p in EXCHANGE)
+    var.SILENCED.update(p.nick for p in SILENCE)
+    var.LYCANTHROPES.update(p.nick for p in LYCANTHROPY)
+    # pestilence doesn't take effect on immunized players
+    var.DISEASED.update({p.nick for p in PESTILENCE} - var.IMMUNIZED)
+    var.LUCKY.update(p.nick for p in LUCK)
+    var.MISDIRECTED.update(p.nick for p in MISDIRECTION)
+
+@event_listener("abstain")
+def on_abstain(evt, var, user):
+    if user in NARCOLEPSY:
+        user.send(messages["totem_narcolepsy"])
+        evt.prevent_default = True
+
+@event_listener("lynch")
+def on_lynch(evt, var, user):
+    if user in NARCOLEPSY:
+        user.send(messages["totem_narcolepsy"])
+        evt.prevent_default = True
+
+@event_listener("assassinate")
+def on_assassinate(evt, var, killer, target, prot):
+    if prot == "totem":
+        var.ACTIVE_PROTECTIONS[target.nick].remove("totem")
+        evt.prevent_default = True
+        evt.stop_processing = True
+        channels.Main.send(messages[evt.params.message_prefix + "totem"].format(killer, target))
```

Comment:
Should figure out what message_prefixes we could be passed and add a comment above this listing the full message keys for each combination. That way when we see a message in en.json we can figure out where it's being used by doing a grep for the key.